src/features/worker-utils/wasi-utils.ts

```diff
@@ -0,0 +1,32 @@
+/**
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export function makeEverythingElseThrow(obj: { [x: string]: {} }): {} {
+  return new Proxy(obj, {
+    get(target, prop: string) {
+      if (prop in target) {
+        return target[prop];
+      }
+      return () => {
+        throw Error(`${prop} not implemented`);
+      };
+    },
+  });
+}
+
+export function makeWasiEnv() {
+  return {
+    environ_sizes_get: () => 0,
+    environ_get: () => 0,
+  };
```

Comment:
Nice.. Is this really all the wasi calls that this code depends on?   I guess you are just doing pure compute.

I'm curious, when you build with emscripten do you see more dependencies than this?   I would hope not.

Reply:
@sbc100 Well, there are some more WASI imports even in this binary, but they are unused:

<details><pre>
Import[8]:
 - func[0] sig=0 <wasi_snapshot_preview1.proc_exit> <- wasi_snapshot_preview1.proc_exit
 - func[1] sig=4 <wasi_snapshot_preview1.environ_sizes_get> <- wasi_snapshot_preview1.environ_sizes_get
 - func[2] sig=4 <wasi_snapshot_preview1.environ_get> <- wasi_snapshot_preview1.environ_get
 - func[3] sig=19 <wasi_snapshot_preview1.fd_seek> <- wasi_snapshot_preview1.fd_seek
 - func[4] sig=9 <wasi_snapshot_preview1.fd_write> <- wasi_snapshot_preview1.fd_write
 - func[5] sig=1 <wasi_snapshot_preview1.fd_close> <- wasi_snapshot_preview1.fd_close
 - func[6] sig=4 <wasi_snapshot_preview1.fd_prestat_get> <- wasi_snapshot_preview1.fd_prestat_get
 - func[7] sig=2 <wasi_snapshot_preview1.fd_prestat_dir_name> <- wasi_snapshot_preview1.fd_prestat_dir_name
</pre></details>

With Emscripten using `-s STANDALONE_WASM=1`, I see this (couldn’t be bothered to compile a version without embind):

<details><pre>
Import[24]:
 - func[0] sig=20 <env._embind_register_value_object_field> <- env._embind_register_value_object_field
 - func[1] sig=4 <env._embind_register_memory_view> <- env._embind_register_memory_view
 - func[2] sig=6 <env._embind_register_integer> <- env._embind_register_integer
 - func[3] sig=4 <env._embind_register_std_wstring> <- env._embind_register_std_wstring
 - func[4] sig=9 <env._embind_register_function> <- env._embind_register_function
 - func[5] sig=0 <env._emval_decref> <- env._emval_decref
 - func[6] sig=11 <wasi_snapshot_preview1.fd_write> <- wasi_snapshot_preview1.fd_write
 - func[7] sig=4 <env._embind_register_float> <- env._embind_register_float
 - func[8] sig=3 <env._embind_register_std_string> <- env._embind_register_std_string
 - func[9] sig=3 <env._embind_register_void> <- env._embind_register_void
 - func[10] sig=0 <env._emval_incref> <- env._emval_incref
 - func[11] sig=0 <env._embind_finalize_value_object> <- env._embind_finalize_value_object
 - func[12] sig=9 <env._embind_register_value_object> <- env._embind_register_value_object
 - func[13] sig=11 <env._emval_new> <- env._emval_new
 - func[14] sig=0 <env.emscripten_notify_memory_growth> <- env.emscripten_notify_memory_growth
 - func[15] sig=0 <wasi_snapshot_preview1.proc_exit> <- wasi_snapshot_preview1.proc_exit
 - func[16] sig=5 <wasi_snapshot_preview1.environ_get> <- wasi_snapshot_preview1.environ_get
 - func[17] sig=5 <wasi_snapshot_preview1.environ_sizes_get> <- wasi_snapshot_preview1.environ_sizes_get
 - func[18] sig=1 <wasi_snapshot_preview1.fd_close> <- wasi_snapshot_preview1.fd_close
 - func[19] sig=27 <wasi_snapshot_preview1.fd_seek> <- wasi_snapshot_preview1.fd_seek
 - func[20] sig=3 <env._embind_register_emval> <- env._embind_register_emval
 - func[21] sig=6 <env._embind_register_bool> <- env._embind_register_bool
 - func[22] sig=1 <env._emval_get_global> <- env._emval_get_global
 - func[23] sig=2 <env.__cxa_thread_atexit> <- env.__cxa_thread_atexit
</pre></details>

Reply:
> @sbc100 Well, there are some more WASI imports even in this binary, but they are unused:

Doesn't this imply there is a bunch of dead code in there?   I imagine that are building here is really what WASI calls a reactor.   Its still pretty experimental but you could try building with `-mexec-model=reactor`.   It works kind of like the `--no-entry` emscripten flag (which you should be using with emscripten if you are not already).    This should eliminate a bunch of those imports I hope (and remove a bunch of dead code).

> 
> With Emscripten using `-s STANDALONE_WASM=1`, I see similar WASI imports, some Emscripten specific stuff + lots of embind (couldn’t be bothered to compile a version without embind):

Reply:
Tried `-mexec-model=reactor`, the imports remain. I guess there are just some dynamic code branches that we never take but that can’t be fully eliminated.